### ISSUE TITLE
Startle NNI is not stopping properly

### ISSUE THREAD

**Issue by Galaxeee** [raphael-group/startle#2]
Hi there,

I tried out Startle-NNI and was able to get the newick outputs. However, it seems like it is running longer iterations than my input and some processes are not closing even if I get the newick output. I wonder if you have encountered this before. Here is the command I was using:
`python src/nni/startle.py -m "$mutation_table_dir" --iterations 5000 --threads 8\
        "$seed_dir" "$mut_dir" \
        --output "$newick_dir" > "$outlog_dir" 2>&1
`
I can also provide the files that I used for the inputs as well as the log file from one of the runs:
- NJ tree(seed): [NJ_256_mu_0.1_pd_1_run_1.txt](https://github.com/raphael-group/startle/files/12689577/NJ_256_mu_0.1_pd_1_run_1.txt)
- lineage data(mut): [mut_256_mu_0.1_pd_1_run_1.csv](https://github.com/raphael-group/startle/files/12689562/mut_256_mu_0.1_pd_1_run_1.csv)
- mutation table: [mut_prob.csv](https://github.com/raphael-group/startle/files/12689601/mut_prob.csv)
- log: [log_startle_256_mu_0.1_pd_1_run_1.txt](https://github.com/raphael-group/startle/files/12689591/log_startle_256_mu_0.1_pd_1_run_1.txt)

I want to test Startle with other methods in terms of running time, but with the current program, it seems that all the runs with different sample sizes have very similar running times and are very slow (~12 hours). It would be a great help if you could look into this.

Best,
Xinhai

**Comment by schmidt73** [comment]
Xinhai,

I am very sorry for the delayed response. I am not sure why the tool is leaving zombie processes, but I think it has to do with Python's multiprocessing module. We will look into it.

As to the runtime issue, there are two things which will significantly help improve the runtime.

First, the number of iterations does not need to be very large. We recommend setting it somewhere between 100-300.

Second, there is a lot of redundancy in your input lineage tracing data. For example, many cells have identical cell-state vectors, and we cannot distinguish the cells based on these vectors. Previously, we had resolved this problem by collapsing cells with identical cell-state vectors, and then reattaching them afterwards, as the optimal solution always places cells with identical cell-state vectors on top of each other (proof is in manuscript). We performed this collapsing with two scripts [prune.py](https://github.com/raphael-group/startle/blob/main/src/nni/prune.py) and [unprune.py](https://github.com/raphael-group/startle/blob/main/src/nni/prune.py), which were run before and after [startle.py](https://github.com/raphael-group/startle/blob/main/src/nni/startle.py). However, running two extra scripts is inconvenient. As such, we have updated the script such that this pruning is done automatically. On the example you gave, it reduces the size of the tree from 510 nodes to 152 nodes.

We hope this helps. If you have any further questions, feel free to respond to this issue/open a new issue. Alternatively, you are welcome to send me an email at henri.schmidt@princeton.edu for a more expedient response -- I rarely check Github issues.
